reactor-core/src/main/java/reactor/util/concurrent/MpscLinkedQueue.java

```diff
@@ -0,0 +1,273 @@
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * The code was inspired by the similarly named JCTools class:
+ * https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/atomic
+ */
+package reactor.util.concurrent;
+
+import java.util.AbstractQueue;
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.BiPredicate;
+
+import reactor.util.annotation.Nullable;
+
+/**
+ * A multi-producer single consumer unbounded queue.
+ * @param <E> the contained value type
+ */
+public class MpscLinkedQueue<E> extends AbstractQueue<E> implements BiPredicate<E, E> {
+	private volatile LinkedQueueNode<E> producerNode;
+
+	private final static AtomicReferenceFieldUpdater<MpscLinkedQueue, LinkedQueueNode> PRODUCER_NODE_UPDATER
+			= AtomicReferenceFieldUpdater.newUpdater(MpscLinkedQueue.class, LinkedQueueNode.class, "producerNode");
+
+	private volatile LinkedQueueNode<E> consumerNode;
+	private final static AtomicReferenceFieldUpdater<MpscLinkedQueue, LinkedQueueNode> CONSUMER_NODE_UPDATER
+			= AtomicReferenceFieldUpdater.newUpdater(MpscLinkedQueue.class, LinkedQueueNode.class, "consumerNode");
+
+	public MpscLinkedQueue() {
+		LinkedQueueNode<E> node = new LinkedQueueNode<>();
+		CONSUMER_NODE_UPDATER.lazySet(this, node);
+		PRODUCER_NODE_UPDATER.getAndSet(this, node);// this ensures correct construction:
+		// StoreLoad
+	}
+
+
+	/**
+	 * {@inheritDoc} <br>
+	 * <p>
+	 * IMPLEMENTATION NOTES:<br>
+	 * Offer is allowed from multiple threads.<br>
+	 * Offer allocates a new node and:
+	 * <ol>
+	 * <li>Swaps it atomically with current producer node (only one producer 'wins')
+	 * <li>Sets the new node as the node following from the swapped producer node
+	 * </ol>
+	 * This works because each producer is guaranteed to 'plant' a new node and link the old node. No 2
+	 * producers can get the same producer node as part of XCHG guarantee.
+	 *
+	 * @see java.util.Queue#offer(java.lang.Object)
+	 */
+	@Override
+	@SuppressWarnings("unchecked")
+	public final boolean offer(@Nullable final E e) {
+		Objects.requireNonNull(e);
+
+		final LinkedQueueNode<E> nextNode = new LinkedQueueNode<>(e);
+		final LinkedQueueNode<E> prevProducerNode = PRODUCER_NODE_UPDATER.getAndSet(this, nextNode);
+		// Should a producer thread get interrupted here the chain WILL be broken until that thread is resumed
+		// and completes the store in prev.next.
+		prevProducerNode.soNext(nextNode); // StoreStore
+		return true;
+	}
+
+
+	@Override
```

Comment:
I think it would be great for the javadoc to be overridden to, to make it explicit this is an atomic offering of two elements and not just a regular test.